Fix some deprecation warnings

cbliard opened the pull request:
# Ticket

* https://community.openproject.org/wp/63574 (one part of the PR)

# What are you trying to accomplish?

Fix some annoying deprecation warnings visible when running tests:

```
DEPRECATION WARNING: `to_time` will always preserve the full timezone rather than offset of the receiver in Rails 8.1. To opt in to the new behavior, set `config.active_support.to_time_preserves_timezone = :zone`. (called from <top (required)> at /Users/cbliard/code/opf/openproject/dev/config/environment.rb:35)
DEPRECATION WARNING: `config.active_job.enqueue_after_transaction_commit` is deprecated and will be removed in Rails 8.1. This configuration can still be set on individual jobs using `self.enqueue_after_transaction_commit=`, but due the nature of this behavior, it is not recommended to be set globally. (called from <top (required)> at /Users/cbliard/code/opf/openproject/dev/app/mailers/application_mailer.rb:31)
/Users/cbliard/.local/share/mise/installs/ruby/3.4.5/lib/ruby/gems/3.4.0/gems/grape-2.4.0/lib/grape/util/registry.rb:10:in 'Grape::Util::Registry#register': json is already registered with class Bim::Bcf::API::ErrorFormatter::Json (StructuredWarnings::StandardWarning)
```

I was not able to remove the one from Grape.

# What approach did you choose and why?

Actually we were already using the recommended defaults:
- for `config.active_job.enqueue_after_transaction_commit`, default is `false` and we were using `:never` which is equivalent to `false`
- for `config.active_support.to_time_preserves_timezone` we were using `:zone` but too late in the loading process. Moving it to `config/application.rb` fixed the warning.
- for grape, the error formatter for json gets registered automatically when it inherits `Grape::ErrorFormatter::Base`. So the default formatter `Grape::ErrorFormatter::Json` is registered, and then when we define ours, `Bim::Bcf::API::ErrorFormatter::Json`, it registers it and it overrides the default one. That outputs a warning by default (with a wrong message by the way). Not sure if we can avoid it...

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
Fix deprecation warning about `enqueue_after_transaction_commit`

Warning was:
> DEPRECATION WARNING:
> `config.active_job.enqueue_after_transaction_commit` is deprecated and
> will be removed in Rails 8.1. This configuration can still be set on
> individual jobs using `self.enqueue_after_transaction_commit=`, but
> due the nature of this behavior, it is not recommended to be set
> globally.

Actually, using `:never` sets it to `false`, and `false` is the default
value. So removing this setting keeps the same behavior and avoids the
deprecation warning.

cbliard committed:
Fix deprecation warning for `to_time`

Warning was:

> DEPRECATION WARNING: `to_time` will always preserve the full timezone
> rather than offset of the receiver in Rails 8.1. To opt in to the new
> behavior, set `config.active_support.to_time_preserves_timezone =
> :zone`.

Actually we were already using `:zone`, but it was set in a file
required late in the loading process. Moving it to
`config/application.rb` fixes the deprecation warning.

myabc reviewed (approved):
Awesome! thanks for doing this @cbliard!

With regards to the `Grape::Util::Registry` warning, I took a brief look myself. I thought renaming the custom formatters might do the trick, but unfortunately it didn't resolve the warning.

```ruby
      formatter :bcf_json, ::API::Formatter.new

      default_format :bcf_json

      error_representer ::Bim::Bcf::API::V2_1::Errors::ErrorRepresenter, "application/json; charset=utf-8"
      error_formatter :bcf_json, ::Bim::Bcf::API::ErrorFormatter::Json
      default_error_formatter :bcf_json
```

cbliard commented:
> With regards to the `Grape::Util::Registry` warning, I took a brief look myself. I thought renaming the custom formatters might do the trick, but unfortunately it didn't resolve the warning.
> 
> ```ruby
>       formatter :bcf_json, ::API::Formatter.new
> 
>       default_format :bcf_json
> 
>       error_representer ::Bim::Bcf::API::V2_1::Errors::ErrorRepresenter, "application/json; charset=utf-8"
>       error_formatter :bcf_json, ::Bim::Bcf::API::ErrorFormatter::Json
>       default_error_formatter :bcf_json
> ```

Aha, good point! I think you're in the right direction: the error formatter is automatically registered using its name, so if it's named `::Bim::Bcf::API::ErrorFormatter::BcfJson`, it will automatically register as `:bcf_json`.

Could you try like this @myabc ?

myabc committed:
Fix Grape Registry warning due to naming collision

Renames `error_formatter` declared in `BCF::API::Root` and its class.
This should resolve the following warning on booting Rails:

    /home/.gem/ruby/3.4.5/gems/grape-2.4.0/lib/grape/util/registry.rb:10:in 'Grape::Util::Registry#register': json is already registered with class Bim::Bcf::API::ErrorFormatter::Json (StructuredWarnings::StandardWarning)

myabc commented:
> Aha, good point! I think you're in the right direction: the error formatter is automatically registered using its name, so if it's named `::Bim::Bcf::API::ErrorFormatter::BcfJson`, it will automatically register as `:bcf_json`.
> 
> Could you try like this @myabc ?

@cbliard great - that worked. I've pushed the fix.

myabc commented on modules/bim/app/controllers/bim/bcf/api/root.rb:
```diff
@@ -39,7 +39,7 @@ class Root < ::API::RootAPI
 
       error_representer ::Bim::Bcf::API::V2_1::Errors::ErrorRepresenter, "application/json; charset=utf-8"
-      error_formatter :json, ::Bim::Bcf::API::ErrorFormatter::Json
+      error_formatter :bcf_json, ::Bim::Bcf::API::ErrorFormatter::BcfJson
```
Not sure if we also need to call `default_error_formatter`. ~~My local BIM environment is stale~~, so I'll see what CI says...

NobodysNightmare replied:
I'd have expected that the class needs to be renamed, but that we remain at `:json` as the symbol name for the formatter.

I considered the default behaviour to be a bug in Grape, but they seem to think that's not the case: https://github.com/ruby-grape/grape/issues/2591

NobodysNightmare commented:
FYI I added a reference to a ticket that this PR will close.

cbliard commented:
The failing test `./spec/features/activities/work_package/activities_spec.rb:678` is passing locally and I don't manage to reproduce the failure locally.
Merging.
